common/djangoapps/track/eventshim.py

```diff
@@ -0,0 +1,380 @@
+"""
+EventShims that create
+"""
+
+# pylint: disable=missing-docstring
+
+import json
+import logging
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import UsageKey
+
+log = logging.getLogger(__name__)
+
+
+class DottedPathMapping(object):
+    """
+    Dictionary-like object for creating keys of dotted paths.
+
+    If a key is created that ends with a dot, it will be treated as a path
+    prefix.  Any value whose prefix matches the dotted path can be used
+    as a key for that value, but only the most specific match will
+    be used.
+    """
+
+    def __init__(self, registry=None):
+        self._match_registry = {}
+        self._prefix_registry = {}
+        self.update(registry or {})
+
+    def __contains__(self, key):
+        try:
+            _ = self[key]
+            return True
+        except KeyError:
+            return False
+
+    def __getitem__(self, key):
+        if key in self._match_registry:
+            return self._match_registry[key]
+        if isinstance(key, basestring):
+            for prefix in sorted(self._prefix_registry, reverse=True):
+                if key.startswith(prefix):
+                    return self._prefix_registry[prefix]
+        raise KeyError('Key {} not found in {}'.format(key, type(self)))
+
+    def __setitem__(self, key, value):
+        if key.endswith('.'):
+            self._prefix_registry[key] = value
+        else:
+            self._match_registry[key] = value
+
+    def __delitem__(self, key):
+        if key.endswith('.'):
+            del self._prefix_registry[key]
+        else:
+            del self._match_registry[key]
+
+    def get(self, key, default=None):
+        try:
+            self[key]
+        except KeyError:
+            return default
+
+    def update(self, dict_):
+        for key, value in dict_:
+            self[key] = value
+
+    def keys(self):
+        return self._match_registry.keys() + self._prefix_registry.keys()
+
+
+class EventShimRegistry(object):
```

Comment:
After discussion on HipChat, we're going with EventTransformers for the EventShims.  Along with that, I'm renaming the Processor to PrefixedEventProcessor (because PrefixedEventTransformerProcessor is (1) too long and (2) reveals implementation details that the caller doesn't need.)